src/secret_sharing/array.rs

```diff
@@ -0,0 +1,245 @@
+    const ZERO: Self = Self([V::ZERO; N]);
+
+    fn capacity() -> usize {
+        N
+    }
+
+    fn index(&self, index: usize) -> V {
+        self.0[index]
+    }
+
+    fn from_item(item: V) -> Self {
+        let mut res = Self::ZERO;
+        res.0[0] = item;
+        res
+    }
+
+    /*
+    fn multiply_elements(lhs: &Self, rhs: &Self) -> Self {
+        StdArray(array::from_fn(|i| lhs.0[i] * rhs.0[i]))
+    }
+    */
+}
+
+impl<V: SharedValue, const N: usize> TryFrom<Vec<V>> for StdArray<V, N> {
+    type Error = ();
+    fn try_from(value: Vec<V>) -> Result<Self, Self::Error> {
+        value.try_into().map(Self).map_err(|_| ())
+    }
+}
+
+impl<'a, 'b, V: SharedValue, const N: usize> Add<&'b StdArray<V, N>> for &'a StdArray<V, N> {
+    type Output = StdArray<V, N>;
+
+    fn add(self, rhs: &'b StdArray<V, N>) -> Self::Output {
+        //StdArray(self.0.iter().zip(rhs.0.iter()).map(|(a, b)| *a + *b).collect::<Vec<_>>().try_into().unwrap())
+        StdArray(array::from_fn(|i| self.0[i] + rhs.0[i]))
+    }
+}
+
+impl<V: SharedValue, const N: usize> Add<Self> for StdArray<V, N> {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Add::add(&self, &rhs)
+    }
+}
+
+// add(owned, ref) should be preferred over this.
+impl<V: SharedValue, const N: usize> Add<StdArray<V, N>> for &StdArray<V, N> {
+    type Output = StdArray<V, N>;
+
+    fn add(self, rhs: StdArray<V, N>) -> Self::Output {
+        Add::add(self, &rhs)
+    }
+}
+
+impl<V: SharedValue, const N: usize> Add<&StdArray<V, N>> for StdArray<V, N> {
+    type Output = Self;
+
+    fn add(self, rhs: &Self) -> Self::Output {
+        Add::add(&self, rhs)
+    }
+}
+
+impl<V: SharedValue, const N: usize> AddAssign<&Self> for StdArray<V, N> {
+    fn add_assign(&mut self, rhs: &Self) {
+        for (a, b) in self.0.iter_mut().zip(rhs.0.iter()) {
+            *a += *b;
+        }
+    }
+}
+
+impl<V: SharedValue, const N: usize> AddAssign<Self> for StdArray<V, N> {
+    fn add_assign(&mut self, rhs: Self) {
+        AddAssign::add_assign(self, &rhs);
+    }
+}
+
+impl<V: SharedValue, const N: usize> Neg for &StdArray<V, N> {
+    type Output = StdArray<V, N>;
+
+    fn neg(self) -> Self::Output {
+        StdArray(
+            self.0
+                .iter()
+                .map(|x| -*x)
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap(),
+        )
+    }
+}
+
+impl<V: SharedValue, const N: usize> Neg for StdArray<V, N> {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        Neg::neg(&self)
+    }
+}
+
+impl<V: SharedValue, const N: usize> Sub<Self> for &StdArray<V, N> {
+    type Output = StdArray<V, N>;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        StdArray(
+            self.0
+                .iter()
+                .zip(rhs.0.iter())
+                .map(|(a, b)| *a - *b)
+                .collect::<Vec<_>>()
```

Comment:
do we need a todo here?

Reply:
I assume to avoid the allocation? Yes, I agree that is a good idea. I need to scrub all of the arithmetic impls. I have already changed a few to use `array::from_fn` instead of allocating.